src/stan/io/json/json_data_handler.hpp

```diff
@@ -7,174 +7,525 @@
+  std::map<std::string, array_dims> slot_dims_map;
+  std::map<std::string, bool> int_slots_map;
+  std::vector<double> values_r;  // accumulates real var values
+  std::vector<int> values_i;     // accumulates int var values
+  size_t array_start_i;          // index into values_i
+  size_t array_start_r;          // index into values_r
+  int event;
+
+  void reset_values() {
+    // Once var values have been copied into var_context maps,
+    // clear the accumulator vectors.
+    values_r.clear();
+    values_i.clear();
+    array_start_i = 0;
+    array_start_r = 0;
+  }
+
+  inline std::string key_str() {
+    return boost::algorithm::join(key_stack, ".");
+  }
+
+  std::string outer_key_str() {
+    std::string result;
+    if (key_stack.size() > 1) {
+      std::string slot = key_stack.back();
+      key_stack.pop_back();
+      result = key_str();
+      key_stack.push_back(slot);
+    }
+    return result;
   }
 
   bool is_init() {
-    return (key_.size() == 0 && values_r_.size() == 0 && values_i_.size() == 0
-            && dims_.size() == 0 && dims_verify_.size() == 0
-            && dims_unknown_.size() == 0 && dim_idx_ == 0 && dim_last_ == 0
-            && is_int_);
+    return (key_stack.empty() && var_types_map.empty() && slot_types_map.empty()
+            && values_r.empty() && values_i.empty() && slot_dims_map.empty()
+            && array_start_i == 0 && array_start_r == 0
+            && int_slots_map.empty());
+  }
+
+  bool is_array_tuples(const std::vector<std::string>& keys) {
+    std::vector<std::string> stack(keys);
+    std::string key;
+    stack.pop_back();
+    while (!stack.empty()) {
+      key = boost::algorithm::join(stack, ".");
+      if (slot_types_map[key] == meta_type::ARRAY_OF_TUPLES)
+        return true;
+      stack.pop_back();
+    }
+    return false;
+  }
+
+  array_dims get_outer_dims(const std::vector<std::string>& keys) {
+    std::vector<std::string> stack(keys);
+    std::string key;
+    stack.pop_back();
+    while (!stack.empty()) {
+      key = boost::algorithm::join(stack, ".");
+      if (slot_dims_map.count(key) == 1)
+        return slot_dims_map[key];
+      stack.pop_back();
+    }
+    key = boost::algorithm::join(keys, ".");
+    if (slot_dims_map.count(key) != 1)
+      unexpected_error(key);
+    return slot_dims_map[key];
+  }
+
+  void set_outer_dims(array_dims update) {
+    std::vector<std::string> stack = key_stack;
+    std::string key;
+    stack.pop_back();
+    while (!stack.empty()) {
+      key = boost::algorithm::join(stack, ".");
+      if (slot_dims_map.count(key) == 1)
+        break;
+      stack.pop_back();
+    }
+    if (stack.empty()) {
+      key = boost::algorithm::join(key_stack, ".");
+      unexpected_error(key);
+    }
+    slot_dims_map[key] = update;
+  }
+
+  void promote_to_double() {
+    if (int_slots_map[key_str()]) {
+      int_slots_map[key_str()] = false;
+      values_r.reserve(values_i.size());
```

Comment:
Is values_r always empty at this point? If not, we need to `reserve(values_r.size() + values_i.size())` (reserve is for total size, not additional size)

Reply:
yes, `values_r` is always empty.   we're just processing one slot's worth of data here.